flooddrake/adaptive_timestepping.py

```diff
@@ -0,0 +1,111 @@
+
+        self.c_w_s = Function(FunctionSpace(self.mesh, 'DG', 0))
+
+        self.max_cell_length = MinDx(self.V.mesh())
+
+        self.min_wave_speed_kernel_2d = """ const double g=9.8; float min=10000000, wave_speed=0; int a=0;
+        for(int i=0;i<vert_u_cell.dofs;i++){
+            if (vert_cell[i][0]<=0){
+                wave_speed=10000000;
+                a=a+1;
+            }
+            if (vert_cell[i][0]>0){
+                wave_speed=sqrt(pow((vert_u_cell[i][0]/vert_cell[i][0]),2)+pow((vert_v_cell[i][0]/vert_cell[i][0]),2))+sqrt(g*vert_cell[i][0]);
+            }
+            min=fmin(wave_speed,min);
+        }
+        if (a==vert_u_cell.dofs){
+            cell_wave_speed[0][0]=10000000;
+        }
+        if (a<vert_u_cell.dofs){
+            cell_wave_speed[0][0]=1/min;
+        }
+        """
+
+        self.min_wave_speed_kernel_1d = """ const double g=9.8; float min=10000000, wave_speed=0; int a=0;
+        #define ABS(X) (((X) < 0) ? (-X) : (X))
+        for(int i=0;i<vert_u_cell.dofs;i++){
+            if (vert_cell[i][0]<=0){
+                wave_speed=10000000;
+                a=a+1;
+            }
+            if (vert_cell[i][0]>0){
+                wave_speed=ABS(vert_u_cell[i][0]/vert_cell[i][0])+sqrt(g*vert_cell[i][0]);
+            }
+            min=fmin(wave_speed,min);
+        }
+        if (a==vert_u_cell.dofs){
+            cell_wave_speed[0][0]=10000000;
+        }
+        if (a<vert_u_cell.dofs){
+            cell_wave_speed[0][0]=1/min;
+        }
+        """
+
+        super(AdaptiveTimestepping, self).__init__()
+
+    def FindTimestep(self, w):
+        """ Finds the CFL criterion satisfying timestep of the DG flooddrake model
+
+            :param w: state vector
+
+        """
+
+        if self.V.mesh().geometric_dimension() == 2:
+
+            # split functions
+            h, mu, mv = split(w)
+
+        if self.V.mesh().geometric_dimension() == 1:
+
+            # split functions
+            h, mu = split(w)
+
+        # find minimum cfl timestep
+        self.c_w_s.assign(0)
+
+        if self.V.mesh().geometric_dimension() == 2:
+
+            par_loop(self.min_wave_speed_kernel_2d, dx, {"cell_wave_speed": (self.c_w_s, RW),
+                                                         "vert_cell": (h, READ),
+                                                         "vert_u_cell": (mu, READ),
+                                                         "vert_v_cell": (mv, READ)})
+
+        if self.V.mesh().geometric_dimension() == 1:
+
+            par_loop(self.min_wave_speed_kernel_1d, dx, {"cell_wave_speed": (self.c_w_s, RW),
+                                                         "vert_cell": (h, READ),
+                                                         "vert_u_cell": (mu, READ)})
+
+        cfl_timestep = self.max_cell_length * self.c_w_s.comm.allreduce(self.c_w_s.dat.data_ro.min(),
+                                                                        MPI.MIN)
```

Comment:
I'm confused, I would have thought that the CFL constraint comes from the the largest wave speed and the smallest cell, not the smallest wave speed and the largest cell.

Moreover, is this not overly conservative.  You presumably have a cell-wise CFL constraint.  So why not compute the CFL on each cell and take the appropriate min/max over that.  Rather than taking a min over cell sizes and a min over wave speeds.